boards: arm: Introduce Infineon CYW920829M2EVK-02 board

sreeramIfx opened the pull request:
- Add initial version of CYW920829M2EVK-02 board.
- [drivers: clock_control] Make it possible to set up both iho and imo clocks instead of just one or the other

npal-cy commented on soc/infineon/cat1b/cyw20829/soc.c:
```diff
@@ -0,0 +1,75 @@
+ */
+
+/**
+ * @brief Infineon PSOC EDGE84 soc.
```
Update `brief` it should be CYW920829 soc

nordicjm commented on boards/infineon/cyw920829m2evk_02/Kconfig.defconfig:
```diff
@@ -0,0 +1,4 @@
+# The Infineon AIROC™ CYW20829 Bluetooth® LE evaluation kit (CYW92089M2EVK-02)
+
+# Copyright (c) 2024 Cypress Semiconductor Corporation.
```
delete empty file

sreeramIfx replied:
removed..

nordicjm commented on boards/infineon/cyw920829m2evk_02/doc/img/board.jpg:
Convert to webp and put through https://tinypng.com/

sreeramIfx replied:
uploaded the modified image and updated the index.rst reference

nordicjm commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -55,6 +62,11 @@ zephyr_library_sources(${pdl_drv_dir}/source/cy_ipc_pipe.c)
 zephyr_library_sources(${pdl_drv_dir}/source/cy_ipc_sema.c)
 zephyr_library_sources(${pdl_drv_dir}/source/cy_prot.c)
 zephyr_library_sources(${pdl_drv_dir}/source/cy_sysclk.c)
+if (CONFIG_SOC_FAMILY_INFINEON_CAT1B)
```
```suggestion
if(CONFIG_SOC_FAMILY_INFINEON_CAT1B)
```

sreeramIfx replied:
done

nordicjm commented on modules/hal_infineon/mtb-template-cat1/CMakeLists.txt:
```diff
@@ -5,10 +5,22 @@
+set(cat1b_dir ${template_dir}/files/templates/cat1b)
 
-zephyr_include_directories(${cat1a_dir}/COMPONENT_MTB)
+if (CONFIG_SOC_FAMILY_INFINEON_CAT1A)
```
```suggestion
if(CONFIG_SOC_FAMILY_INFINEON_CAT1A)
```

nordicjm commented on modules/hal_infineon/mtb-template-cat1/CMakeLists.txt:
```diff
@@ -5,10 +5,22 @@
+
+
+# Add support cyw20829 (cat1b)
+if (CONFIG_SOC_FAMILY_INFINEON_CAT1B)
```
```suggestion
if(CONFIG_SOC_FAMILY_INFINEON_CAT1B)
```

nordicjm commented on soc/infineon/cat1b/Kconfig:
```diff
@@ -0,0 +1,20 @@
+
+# Series definitions
+config SOC_SERIES_CYW20829
+	bool
```
```suggestion
```

nordicjm commented on soc/infineon/cat1b/Kconfig:
```diff
@@ -0,0 +1,20 @@
+
+# Infineon CAT1B devices
+
+# Family definitions
+config SOC_FAMILY_INFINEON_CAT1
+	bool
```
family goes in `Kconfig.soc`

sreeramIfx replied:
updated

nordicjm commented on soc/infineon/cat1b/Kconfig.defconfig:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2024 Cypress Semiconductor Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+rsource "*/Kconfig.defconfig"
```
missing if guard

sreeramIfx replied:
updated

nordicjm commented on arch/Kconfig:
```diff
@@ -866,6 +866,13 @@ config CODE_DATA_RELOCATION
 	  the target regions should be specified in CMakeLists.txt using
 	  zephyr_code_relocate().
 
+config SKIP_ROM_START_SNIPPET_LD
+	bool
+	help
+	  Configuration to skip including snippets-rom-start.ld to
+	  linker script. It can be used for platforms that require custom
+	  ROM START section or specific location for this section.
```
@tejlmand

npal-cy commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+   west debug --openocd path/to/infineon/openocd/bin/openocd
+
+Once the gdb console starts after executing the west debug command, you may
+now set breakpoints and perform other standard GDB debugging on the PSoC 6 CM4 core.
```
replace `PSoC 6 CM4 core`  to `CYW20829 device`

ifyall commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+Hardware
+********
+
+For more information about the PSoC 62 MCU SoC and CYW920829M2EVK-02 board:
```
Replace PSoC 62 MCU SoC with CYW20829 SOC

ifyall commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+
+To get the OpenOCD package, it is required that you
+
+1. Download the software ModusToolbox 3.1. https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
```
Might as well change this to ModusToolbox 3.2, since that came out a few weeks ago. Another option is to say ModusToolbox 3.1 or greater, so we don't have to change when new releases come out.

ifyall commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+
+1. Download the software ModusToolbox 3.1. https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
+2. Once downloaded add the path to access the Scripts folder provided by ModusToolbox
+   export PATH=$PATH:/path/to/ModusToolbox/tools_3.1/openocd/scripts
```
We should note here that tools_x.y depends on the version of ModusToolbox you have.

ifyall commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+Programming and Debugging
+*************************
+
+The CYW920829M2EVK-02 includes an onboard programmer/debugger (KitProg2) with
```
Check this, I think its KitProg3, not 2.

ifyall commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -6,16 +6,23 @@
+zephyr_library_sources_ifdef(CONFIG_SOC_FAMILY_INFINEON_CAT1A ${pdl_dev_cat1a_dir}/source/cy_device.c)
+
+
+zephyr_include_directories(${pdl_drv_dir}/include)
```
This is a duplicate of line 15, so not necessary.

ifyall commented on soc/infineon/cat1b/cyw20829/soc.c:
```diff
@@ -0,0 +1,75 @@
+#include <system_cat1b.h>
+#include "cy_pdl.h"
+
+#define CY_IPC_MAX_ENDPOINTS (8UL)
```
define appears to be unused. Remove it.

nordicjm commented on soc/infineon/cat1b/Kconfig.soc:
```diff
@@ -0,0 +1,17 @@
+	bool
+
+# SOC series
+config SOC_SERIES_CYW20829
+	bool
```
Move to the soc Kconfig.soc instead of being at the parents

nordicjm commented on soc/infineon/cat1b/cyw20829/scripts/generate_app_header.py:
```diff
@@ -0,0 +1,37 @@
+
+boot_strap_size = int(sys.argv[2], 16).to_bytes(4, 'little')
+
+ram_end_addr = 0x20040000
```
shouldn't this come from dts?

nordicjm commented on boards/infineon/cyw920829m2evk_02/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+find_program(SREC_CAT srec_cat)
+if(NOT SREC_CAT)
+    message(WARNING "srec_cat not found!")
```
cmake indent is 2 spaces

nordicjm commented on boards/infineon/cyw920829m2evk_02/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+if(NOT SREC_CAT)
+    message(WARNING "srec_cat not found!")
+else()
+	set(generate_app_header ${ZEPHYR_BASE}/soc/infineon/cat1b/cyw20829/scripts/generate_app_header.py)
+	set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+		COMMAND ${PYTHON_EXECUTABLE}
+		${generate_app_header}
+		${ZEPHYR_BINARY_DIR}/app_header.bin
+		0x2400
+		COMMAND ${SREC_CAT}
+		${ZEPHYR_BINARY_DIR}/app_header.bin -Binary -offset 0x60000000
+		${ZEPHYR_BINARY_DIR}/zephyr.bin -Binary -offset 0x60000050
+		-o ${ZEPHYR_BINARY_DIR}/zephyr.hex -Intel -Output_Block_Size=16
+	)
```
shouldn't this be done at the soc level?

tejlmand commented on soc/infineon/cat1b/Kconfig.soc:
```diff
@@ -0,0 +1,13 @@
+# an affiliate of Cypress Semiconductor Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# Family definitions
+config SOC_FAMILY_INFINEON_CAT1
+	bool
```
this is also defined here: https://github.com/zephyrproject-rtos/zephyr/blob/80a6a206289537fa220e34fe485b11fccef9ac85/soc/cypress/Kconfig.soc#L7-L8

I know Cypress Semi is acquired  by Infenion, but scatter this in two trees is confusing.
Are there any plans to cleanup this ?
I would prefer things to be aligned, at least as long as those SoC are part of same SoC family.

npal-cy replied:
it will cleanup in https://github.com/zephyrproject-rtos/zephyr/pull/70686

tejlmand commented on soc/infineon/cat1b/cyw20829/Kconfig:
```diff
@@ -0,0 +1,15 @@
+	select CPU_CORTEX_M33
+	select CPU_HAS_FPU
+	select DYNAMIC_INTERRUPTS
+	select SOC_FAMILY_INFINEON_CAT1
```
wouldn't it make more sense to have the cat1b family selecting the cat1 family here:
https://github.com/zephyrproject-rtos/zephyr/blob/d59bd090ad4138a3840143b3ffe2cd8ac54f0d40/soc/infineon/cat1b/Kconfig.soc#L9-L10

as:
```
config SOC_FAMILY_INFINEON_CAT1B 
	bool 
	select SOC_FAMILY_INFINEON_CAT1
```

tejlmand commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+zephyr_compile_definitions(COMPONENT_CM33)
+
+# Use custom linker script
+set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/soc/infineon/cat1b/cyw20829/linker.ld CACHE INTERNAL "")
```
you are setting a custom linker script, so why the other changes in Zephyr related to linker scripts ?

npal-cy replied:
it just adds specific bootstrap sections, rest RAM/ROM is used Zephyr default: https://github.com/zephyrproject-rtos/zephyr/pull/70961/files#diff-b8ef5cf2af3dcc18025c6641731242b834cf7fdb021d94a16ab579dd7772d65aR180

see my comment above with details.

tejlmand replied:
thanks for the explanation.

tejlmand reviewed (changes requested):
initial review of build related files.

npal-cy commented on include/zephyr/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -123,11 +123,13 @@ SECTIONS
     SECTION_PROLOGUE(rom_start,,)
 	{
 
+#ifndef CONFIG_SKIP_ROM_START_SNIPPET_LD
```
For CYW920829 we need to have rom_start (vector table, etc)  located in specific bootstrap section of sram (rom boot will care to load bootstrap to sram).

 `SKIP_ROM_START_SNIPPET_LD` allows us to move rom_start section to custom linker script where we declare bootstrap section (https://github.com/zephyrproject-rtos/zephyr/pull/70961/files#diff-b8ef5cf2af3dcc18025c6641731242b834cf7fdb021d94a16ab579dd7772d65aR81) and at same time we continue use Zephyr linker for application (https://github.com/zephyrproject-rtos/zephyr/pull/70961/files#diff-b8ef5cf2af3dcc18025c6641731242b834cf7fdb021d94a16ab579dd7772d65aR180)
 
We really do not want to use fully custom script, but find some way to extend existing one. Let us know If there is some better way to keep rom_start in separate section (bootstrap).

sreeramIfx replied:
I have reverted the addition of SKIP_ROM_START_SNIPPET_LD and added custom linker script to handle this for now.

npal-cy commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+# Generate Application header
+find_program(SREC_CAT srec_cat)
+if(NOT SREC_CAT)
+  message(WARNING "srec_cat not found!")
```
opoenocd will return the error during program not patched hex (if skipped post build..)
we use `WARNING` to unblock CI/build run, in documentation we added section about  srec_cat usage.

hakanjansson replied:
This warning is easily missed and a user risk ending up with a bad hex file without realizing it.

Since the binaries are contiguous, why not use something like `cmake -E cat` followed by `objcopy` instead and remove the need for `srec_cat`?

tejlmand replied:
> opoenocd will return the error during program not patched hex (if skipped post build..)

but will the error at this time make it clear to the user that it is due to a missing `srec_cat` ?

> we use WARNING to unblock CI/build run

sound like the docker image should have `srec_cat` should be included in the docker image ?

sreeramIfx replied:
I think adding srec_cat to the docker image and updating this to be an ERROR so that its more clear to the user makes sense.. 

Actually, looking at the https://github.com/zephyrproject-rtos/docker-image/blob/main/Dockerfile.base srecord is already in the list. Can you please confirm that this is the correct repo?

hakanjansson commented on boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02-common.dtsi:
```diff
@@ -0,0 +1,58 @@
+#include <zephyr/dt-bindings/input/input-event-codes.h>
+
+/ {
+	aliases {
+		led0 = &user_led0;
+		sw0 = &user_bt0;
+	};
```
```suggestion
	aliases {
		led0 = &user_led0;
		sw0 = &user_bt0;
		led1 = &user_led1;
		sw1 = &user_bt1;
	};
```

hakanjansson commented on soc/infineon/cat1b/soc.yml:
```diff
@@ -0,0 +1,12 @@
+family:
+- name: cat1d
```
```suggestion
- name: cat1b
```

hakanjansson commented on dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi:
```diff
@@ -0,0 +1,307 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+	flash0: flash@60002600{
```
This is an external flash and should probably be in the board's DT

sreeramIfx replied:
moved the flash node into the board's dtsi files..

hakanjansson commented on soc/infineon/cat1b/cyw20829/bootstrap.ld:
```diff
@@ -0,0 +1,171 @@
+
+MEMORY
+{
+    BS_CODE (rwx) : ORIGIN = CODE_VMA, LENGTH = CODE_BS_SIZE
+    BS_DATA (rwx) : ORIGIN = DATA_BS_VMA,  LENGTH = DATA_BS_SIZE
```
With the hardcoded configuration above, these regions seems to be identical (=sram_app_header in DT).

sreeramIfx committed:
boards: arm: Introduce Infineon CYW920829M2EVK-02 board

- Add initial version of CYW920829M2EVK-02 board
- [drivers: clock_control] Make it possible to set up both iho and imo
  clocks instead of just one or the other

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,128 @@
+
+Here is an example for the :zephyr:code-sample:`blinky` application.
+
+.. code-block:: console
+
+   cd zephyr
+   west build -p auto -b cyw920829m2evk_02 samples/basic/blinky
```
please use `.. zephyr-app-commands::` (see other boards or code samples)

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,128 @@
+
+To get the OpenOCD package, it is required that you
+
+1. Download the software ModusToolbox 3.2. https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
+2. Once downloaded add the path to access the Scripts folder provided by ModusToolbox
+   export PATH=$PATH:/path/to/ModusToolbox/tools_x.y/openocd/scripts
+3. Add the OpenOCD executable file's path to west flash/debug.
+4. Flash using: west flash --openocd path/to/infineon/openocd/bin/openocd
+5. Debug using: west debug --openocd path/to/infineon/openocd/bin/openocd
```
```suggestion
1. Download the software ModusToolbox 3.2. https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
2. Once downloaded add the path to access the Scripts folder provided by ModusToolbox
   
   .. code-block:: bash

      export PATH=$PATH:/path/to/ModusToolbox/tools_x.y/openocd/scripts

3. Add the OpenOCD executable file's path to west flash/debug.
4. Flash using: ``west flash --openocd path/to/infineon/openocd/bin/openocd``
5. Debug using: ``west debug --openocd path/to/infineon/openocd/bin/openocd``
```

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,128 @@
+
+The CYW920829M2EVK-02 includes an onboard programmer/debugger (KitProg3) to provide debugging, flash programming, and serial communication over USB. Flash and debug commands must be pointed to the Cypress OpenOCD you downloaded above.
+
+On Windows:
```
if you really want to provide instructions both for Windows and Linux then the "OpenOCD installation" section probably needs to be adapted as it seems Linux specific at the moment

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,128 @@
+
+To get the OpenOCD package, it is required that you
+
+1. Download the software ModusToolbox 3.2. https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
```
make this a `` `ModusToolbox 3.2`_ `` link like the other links in this page

kartben reviewed (changes requested):
some doc comments - thanks!

npal-cy commented on soc/infineon/cat1b/cyw20829/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+	default 32
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 48000000
```
could you check that SYS_CLOCK_HW_CYCLES_PER_SEC= 48000000 and  NUM_IRQS = 32 are ok here?

sreeramIfx replied:
fixed the NUM_IRQ

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -0,0 +1,125 @@
+
+On Windows:
+
+.. code-block:: console
+
+   west flash --openocd path/to/infineon/openocd/bin/openocd.exe
+   west debug --openocd path/to/infineon/openocd/bin/openocd.exe
+
+On Linux:
+
+.. code-block:: console
+
+   west flash --openocd path/to/infineon/openocd/bin/openocd
+   west debug --openocd path/to/infineon/openocd/bin/openocd
```
```suggestion
.. code-block:: shell

   west flash --openocd path/to/infineon/openocd/bin/openocd.exe
   west debug --openocd path/to/infineon/openocd/bin/openocd.exe

On Linux:

.. code-block:: shell

   west flash --openocd path/to/infineon/openocd/bin/openocd
   west debug --openocd path/to/infineon/openocd/bin/openocd
```

sreeramIfx replied:
Thank you for the +1, I am working on a PR adding support for 20829 drivers, I will fix this in that PR.

kartben reviewed (approved):
+1 for docs, thanks! (added a minor nit in case this can be addressed should there be further updates to the PR)

tomi-font commented:
FYI this board is provoking build failures on `tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported` because of `cyhal_flash_*()` functions redefined as `cyhal_nvm_*()` ones that don't exist: https://github.com/zephyrproject-rtos/zephyr/actions/runs/9251292259/job/25446729924?pr=73267#step:12:743

sreeramIfx commented:
> FYI this board is provoking build failures on `tests/subsys/mgmt/mcumgr/fs_mgmt_hash_supported` because of `cyhal_flash_*()` functions redefined as `cyhal_nvm_*()` ones that don't exist: https://github.com/zephyrproject-rtos/zephyr/actions/runs/9251292259/job/25446729924?pr=73267#step:12:743

Pushed a fix @ https://github.com/zephyrproject-rtos/zephyr/pull/73377

tomi-font commented:
I should have pointed to the PR I have that does this (among others): https://github.com/zephyrproject-rtos/zephyr/pull/73267
